docs/TPPDialect.md

```diff
@@ -0,0 +1,351 @@
+Non-linear activation functions.
+Types must be the same.
+```mlir
+  tpp.relu ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // RELU op
+  tpp.tanh ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // TANH op
+  tpp.sigmoid ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // SIGMOID op
+  tpp.gelu ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // GELU op
+  tpp.leaky_relu ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // LEAKY_RELU op
+  tpp.elu ins(%0) outs(%1) : memref<MxNxTy> -> memref<MxNxTy> // ELU op
+```
+
+## Quant/Dequant
+Ignore for now.
+Should be as simple as the other unary ones.
+
+## Scatter/Gather
+Ignore for now.
+Should be as simple as the other unary ones.
+
+# Binary Ops
+
+## Element-wise Arith
+Apply element-wise ops..
+Operands of the same type, broadcast by operator.
+If operand's producer is `tpp.broadcast`, add appropriate `BCAST_*` flag to op.
+
+Generic out-of-place syntax:
+```mlir
+  %2 = memref.alloc : memref<NxMxTy>
+  tpp.add ins(%0, %1) outs(%2) : (memref<NxMxTy>, memref<NxMxTy>, memref<NxMxTy>) -> memref<NxMxTy> // ADD
+  tpp.sub ins(%0, %1) outs(%2) : (memref<NxMxTy>, memref<NxMxTy>, memref<NxMxTy>) -> memref<NxMxTy> // SUB
+  tpp.mul ins(%0, %1) outs(%2) : (memref<NxMxTy>, memref<NxMxTy>, memref<NxMxTy>) -> memref<NxMxTy> // MUL
+  tpp.div ins(%0, %1) outs(%2) : (memref<NxMxTy>, memref<NxMxTy>, memref<NxMxTy>) -> memref<NxMxTy> // DIV
+  tpp.muladd ins(%0, %1) outs(%2) : (memref<NxMxTy>, memref<NxMxTy>, memref<NxMxTy>) -> memref<NxMxTy> // MULADD
+```
+
+Depending on the arguments, the operation can also be in-place.
+```mlir
+  // In-place on the first argument, %2 alias %0 -> `a = a + b`, or `a += b`
+  tpp.add ins(%0, %1) outs(%0) : (memref<NxMxTy>, memref<NxMxTy>, memref<NxMxTy>) -> memref<NxMxTy>
+  // In-place on the second argument, %2 alias %1 -> `b = a + b`, or `b += a`
+  tpp.add ins(%0, %1) outs(%1) : (memref<NxMxTy>, memref<NxMxTy>, memref<NxMxTy>) -> memref<NxMxTy>
+  // In-place on the only argument, %2 alias %0 -> `a = a + a`, or `a += a`
+  tpp.add ins(%0, %0) outs(%0) : (memref<NxMxTy>, memref<NxMxTy>, memref<NxMxTy>) -> memref<NxMxTy>
+```
+
+## Matmul
+Matrix-multiply between two memrefs.
+Always out-of-place.
+Operands of the compatible type (M, N, K), broadcast by operator
+If operand's producer is `tpp.broadcast`, add appropriate `BCAST_*` flag to op.
+If operand's producer is `tpp.transpose` and/or `tpp.vnni_pack`, change op to appropriate `MATMUL_X_TRANS_VNNI` variant.
+
+```mlir
+  // C = A x B (new C)
+  %2 = memref.alloc : memref<NxMxTy>
+  tpp.zero(%2) : memref<MxNxTy>
+  tpp.matmul ins(%0, %1) outs(%2) : (memref<NxKxTy>, memref<KxMxTy>) -> memref<NxMxTy>
+```
+
+If the shape of the second argument has rank+1 and the inner dim is 2 or 4, this is a VNNI matmul:
+```mlir
+  tpp.matmul ins(%0, %1) outs(%2) : (memref<64x16xTy>, memref<8x32x2xTy>) -> memref<64x32xTy>
+```
+The output of a VNNI-packed matmul is **not** VNNI packed and needs no _unpacking_.
+
```

Comment:
For future reference, we plan to support in GEMMs at the libxsmm TPP level with C being in either vnni-packed or flat format